lib/input/sftp.go

```diff
@@ -0,0 +1,495 @@
+			),
+			docs.FieldAdvanced(
+				"retry_sleep_duration",
+				"How long it will sleep after failing to connect to the server before trying again, defaults to 5s if not provided.",
+				"10s", "5m",
+			),
+			codec.ReaderDocs,
+			docs.FieldAdvanced("delete_objects", "Whether to delete files from the server once they are processed."),
+		},
+		Categories: []Category{
+			CategoryServices,
+			CategoryNetwork,
+		},
+	}
+}
+
+//------------------------------------------------------------------------------
+
+// SFTPConfig contains configuration fields for the SFTP input type.
+type SFTPConfig struct {
+	Server                string          `json:"server" yaml:"server"`
+	Port                  int             `json:"port" yaml:"port"`
+	Filename              string          `json:"filename" yaml:"filename"`
+	Credentials           SFTPCredentials `json:"credentials" yaml:"credentials"`
+	Path                  string          `json:"path" yaml:"path"`
+	MaxConnectionAttempts int             `json:"max_connection_attempts" yaml:"max_connection_attempts"`
+	RetrySleepDuration    string          `json:"retry_sleep_duration" yaml:"retry_sleep_duration"`
+	Codec                 string          `json:"codec" yaml:"codec"`
+	DeleteObjects         bool            `json:"delete_objects" yaml:"delete_objects"`
+}
+
+// SFTPCredentials contains the credentials for connecting to the SFTP server
+type SFTPCredentials struct {
+	Username string `json:"username" yaml:"username"`
+	Secret   string `json:"secret" yaml:"secret"`
+}
+
+// NewSFTPConfig creates a new SFTPConfig with default values.
+func NewSFTPConfig() SFTPConfig {
+	return SFTPConfig{
+		Server:                "",
+		Port:                  0,
+		Filename:              "",
+		Credentials:           SFTPCredentials{},
+		MaxConnectionAttempts: 10,
+		RetrySleepDuration:    "5s",
+		Path:                  "",
+		Codec:                 "lines",
+		DeleteObjects:         false,
+	}
+}
+
+//------------------------------------------------------------------------------
+
+// SFTP is a benthos reader.Type implementation that reads messages
+// from file(s) on an SFTP server.
+type SFTP struct {
+	conf SFTPConfig
+
+	log   log.Modular
+	stats metrics.Type
+
+	client *sftp.Client
+
+	objectScannerCtor codec.ReaderConstructor
+	keyReader         *sftpTargetReader
+
+	objectMut sync.Mutex
+	object    *sftpPendingObject
+}
+
+// NewSFTP creates a new SFTP input type.
+func NewSFTP(conf SFTPConfig, log log.Modular, stats metrics.Type) (*SFTP, error) {
+	var err error
+	var objectScannerCtor codec.ReaderConstructor
+	if objectScannerCtor, err = codec.GetReader(conf.Codec, codec.NewReaderConfig()); err != nil {
+		return nil, fmt.Errorf("invalid sftp codec: %v", err)
+	}
+
+	s := &SFTP{
+		conf:              conf,
+		log:               log,
+		stats:             stats,
+		objectScannerCtor: objectScannerCtor,
+	}
+
+	err = s.initSFTPConnection()
```

Comment:
Normally the connection of a reader wouldn't be established until `ConnectWithContext` is called, this allows Benthos to manage retries and reconnects as/when needed.